Template parser should ignore comments

```
<div>
  <div class="menu">
    <a class="menu-item" href="" routerLink><div class="logo"></div></a>
    <a class="menu-item" href="about" routerLink>About</a>
    <a class="menu-item" href="param/1" routerLink>Parameter view</a>
    <a class="menu-item" href="resource/" routerLink>Resource view</a>
    <a class="menu-item" href="dynresource" routerLink>Dynamic resource view</a>
  </div>
  <div><!-- Child view goes here --></div>
</div>
```

Cannot be used because of

```
Caused by: java.lang.IllegalArgumentException: Unsupported node type: org.jsoup.nodes.Comment
    at com.vaadin.hummingbird.template.TemplateParser.createBuilder(TemplateParser.java:104)
    at com.vaadin.hummingbird.template.TemplateParser$$Lambda$47/1937301449.apply(Unknown Source)
```

--- a/hummingbird-server/src/main/java/com/vaadin/hummingbird/template/TemplateParser.java
+++ b/hummingbird-server/src/main/java/com/vaadin/hummingbird/template/TemplateParser.java
@@ -17,9 +17,11 @@ package com.vaadin.hummingbird.template;
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.util.Optional;
 
 import org.jsoup.Jsoup;
 import org.jsoup.nodes.Attribute;
+import org.jsoup.nodes.Comment;
 import org.jsoup.nodes.Document;
 import org.jsoup.nodes.Element;
 import org.jsoup.nodes.Node;
@@ -90,16 +92,19 @@ public class TemplateParser {
             }
         }
 
-        TemplateNodeBuilder templateBuilder = createBuilder(children.get(0));
-
-        return templateBuilder.build(null);
+        Optional<TemplateNodeBuilder> templateBuilder = createBuilder(
+                children.get(0));
+        assert templateBuilder.isPresent();
+        return templateBuilder.get().build(null);
     }
 
-    private static TemplateNodeBuilder createBuilder(Node node) {
+    private static Optional<TemplateNodeBuilder> createBuilder(Node node) {
         if (node instanceof Element) {
-            return createElementBuilder((Element) node);
+            return Optional.of(createElementBuilder((Element) node));
         } else if (node instanceof TextNode) {
-            return createTextBuilder((TextNode) node);
+            return Optional.of(createTextBuilder((TextNode) node));
+        } else if (node instanceof Comment) {
+            return Optional.empty();
         } else {
             throw new IllegalArgumentException(
                     "Unsupported node type: " + node.getClass().getName());
@@ -131,6 +136,7 @@ public class TemplateParser {
         element.attributes().forEach(attr -> setBinding(attr, builder));
 
         element.childNodes().stream().map(TemplateParser::createBuilder)
+                .filter(Optional::isPresent).map(Optional::get)
                 .forEach(builder::addChild);
 
         return builder;

--- a/hummingbird-server/src/test/java/com/vaadin/hummingbird/template/TemplateParserTest.java
+++ b/hummingbird-server/src/test/java/com/vaadin/hummingbird/template/TemplateParserTest.java
@@ -137,4 +137,21 @@ public class TemplateParserTest {
         TemplateParser.parse("<div>@child@<span>@child@</span></div>");
     }
 
+    @Test
+    public void parseTopComment() {
+        ElementTemplateNode node = (ElementTemplateNode) TemplateParser
+                .parse("<!-- comment --><div></div>");
+        Assert.assertEquals(0, node.getChildCount());
+        Assert.assertEquals("div", node.getTag());
+    }
+
+    @Test
+    public void parseInnerComment() {
+        ElementTemplateNode node = (ElementTemplateNode) TemplateParser
+                .parse("<div> <!-- comment --> <input> </div>");
+        Assert.assertEquals(4, node.getChildCount());
+        Assert.assertEquals("div", node.getTag());
+        Assert.assertEquals("input",
+                ((ElementTemplateNode) node.getChild(2)).getTag());
+    }
 }

